File: two_pointers/16.3-sum-closest.cpp
/*
 * @lc app=leetcode id=16 lang=cpp
 *
 * [16] 3Sum Closest
 */

#include "bits/stdc++.h"

using namespace std;


// NOTE: three pointers
// @lc code=start
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int diff = INT32_MAX;

        for (int i = 0; i < n && diff!=0; ++i) {
            int l = i + 1;
            int r = n - 1;
            while (l<r)
            {
                int sum = nums[i] + nums[l] + nums[r];
                if (abs(target-sum) < abs(diff)){
                    diff = target - sum;
                }
                if (sum<target){
                    ++l;
                }else{
                    --r;
                }
            }
        }
        return target - diff;
    }
};
// @lc code=end


// NOTE: brute force
class Solution2 {
public:
  int threeSumClosest(vector<int> &nums, int target) {
    int n = nums.size();
    int res = INT32_MAX / 2;
    for (int i = 0; i < n - 2; ++i) {
      for (int j = i + 1; j < n - 1; ++j) {
        for (int k = j + 1; k < n; ++k) {
          int temp = nums[i] + nums[j] + nums[k];
          if (abs(temp - target) < abs(res - target))
            res = temp;
        }
      }
    }
    return res;
  }
};

int main() { cout << INT16_MAX; }

File: two_pointers/16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start
from typing import List


# Time Limit Exceeded
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        res = 10**9
        for i in range(0, n-2):
            for j in range(i+1, n-1):
                for k in range(j+1, n):
                    temp = nums[i]+nums[j]+nums[k]
                    if abs(temp-target) < abs(res-target):
                        res = temp
                        
                        
        return res

# @lc code=end

s = Solution()
print(s.threeSumClosest([0, 0, 0], 1))



